src/server.rs

```diff
@@ -51,35 +71,40 @@ impl DataResponse {
     }
 }
 
-pub async fn metadata_handler(payload: actix_web::web::Bytes) -> actix_web::HttpResponse {
-    let payload_to_str = std::str::from_utf8(&payload);
-    let parse_result = match payload_to_str {
-        Ok(r) => json::parse(r),
-        Err(_err) => Err(json::Error::FailedUtf8Parsing),
-    };
-
-    let body = parse_result.unwrap_or_else(|e| json::object! { "error" => e.to_string() });
+pub async fn metadata_handler(
+    app_state: actix_web::web::Data<context::AppState>,
+    payload: actix_web::web::Json<MetadataMessage>,
+) -> actix_web::HttpResponse {
```

Comment:
so we return `impl Responder` here
```suggestion
) -> impl Responder {
```